connector/new.go

```diff
@@ -0,0 +1,300 @@
+	case providers.Atlassian:
+		conn, err = buildConnector(newAtlassianConnector, parameters, mustAuthenticatedClient, mustWorkspace)
+	case providers.Pipeliner:
+		conn, err = buildConnector(newPipelinerConnector, parameters, mustAuthenticatedClient, mustWorkspace)
+	case providers.Smartlead:
+		conn, err = buildConnector(newSmartleadConnector, parameters, mustAuthenticatedClient)
+	case providers.Marketo:
+		conn, err = buildConnector(newMarketoConnector, parameters, mustAuthenticatedClient, mustWorkspace)
+	case providers.Instantly:
+		conn, err = buildConnector(newInstantlyConnector, parameters, mustAuthenticatedClient)
+	case providers.Apollo:
+		conn, err = buildConnector(newApolloConnector, parameters, mustAuthenticatedClient)
+	case providers.Gong:
+		conn, err = buildConnector(newGongConnector, parameters, mustAuthenticatedClient)
+	case providers.Attio:
+		conn, err = buildConnector(newAttioConnector, parameters, mustAuthenticatedClient)
+	case providers.Pipedrive:
+		conn, err = buildConnector(newPipedriveConnector, parameters, mustAuthenticatedClient)
+	case providers.Zoho:
+		conn, err = buildConnector(newZohoConnector, parameters, mustAuthenticatedClient)
+	case providers.Close:
+		conn, err = buildConnector(newCloseConnector, parameters, mustAuthenticatedClient)
+	case providers.Klaviyo:
+		conn, err = buildConnector(newKlaviyoConnector, parameters, mustAuthenticatedClient)
+	case providers.CustomerJourneysApp:
+		conn, err = buildConnector(newCustomerJourneysAppConnector, parameters, mustAuthenticatedClient)
+	case providers.ConstantContact:
+		conn, err = buildConnector(newConstantContactConnector, parameters, mustAuthenticatedClient)
+	case providers.Keap:
+		conn, err = buildConnector(newKeapConnector, parameters, mustAuthenticatedClient)
+	default:
+		conn, err = newGenericConnector(provider, parameters, mustAuthenticatedClient)
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("error creating %s connector: %w", provider, err)
+	}
+
+	return conn, nil
+}
+
+func buildConnector(
+	builder func(Parameters) (connectors.Connector, error),
+	parameters Parameters,
+	opts ...Option,
+) (connectors.Connector, error) {
+	for _, opt := range opts {
+		opt(&parameters)
+	}
+
+	if parameters.validity.invalid {
+		return nil, parameters.validity.error
+	}
+
+	return builder(parameters)
+}
+
+func newGenericConnector(provider providers.Provider, parameters Parameters, opts ...Option) (connectors.Connector, error) {
+	for _, opt := range opts {
+		opt(&parameters)
+	}
+
+	if parameters.validity.invalid {
+		return nil, parameters.validity.error
+	}
+
+	return generic.NewConnector(
+		provider,
+		generic.WithAuthenticatedClient(parameters.AuthenticatedClient),
+		generic.WithWorkspace(parameters.Workspace),
+	)
+}
+
+func newSalesforceConnector(parameters Parameters) (connectors.Connector, error) {
+	return salesforce.NewConnector(
+		salesforce.WithAuthenticatedClient(parameters.AuthenticatedClient),
```

Comment:
Keeping these options as they are currently, will add a new PR to show how we can reduce the need to have per-package options